lib/spack/spack/spec.py

```diff
@@ -766,13 +767,136 @@ def __str__(self):
+        self.store_by_child = (store_by == EdgeDirection.child)
+
+    def __getitem__(self, key):
+        return self.edges[key]
+
+    def __iter__(self):
+        return iter(self.edges)
+
+    def __len__(self):
+        return len(self.edges)
+
+    def add(self, edge):
+        """Adds a new edge to this object.
+
+        Args:
+            edge (DependencySpec): edge to be added
+        """
+        key = edge.spec.name if self.store_by_child else edge.parent.name
+
+        current_list = self.edges.setdefault(key, [])
+
+        # The list of edges is empty, just append the edge
+        if not current_list:
+            current_list.append(edge)
+            return
+
+        # If there's something in the list, check if we need to update an
+        # already existing entry.
+        for dep in current_list:
+            if edge.spec == dep.spec and edge.parent == dep.parent:
+                dep.add_type(edge.deptypes)
+                break
+        else:
+            current_list.append(edge)
+
+        # Keep the list sorted
+        current_list.sort(key=_sort_by_dep_types)
 
     def __str__(self):
         return "{deps: %s}" % ', '.join(str(d) for d in sorted(self.values()))
 
+    def _cmp_iter(self):
+        for item in sorted(itertools.chain.from_iterable(self.edges.values())):
+            yield item
+
+    def copy(self):
+        """Copies this object and returns a clone"""
+        clone = type(self)()
+        clone.store_by_child = self.store_by_child
+
+        # Copy everything from this dict into it.
+        for dspec in itertools.chain.from_iterable(self.values()):
+            clone.add(dspec.copy())
+
+        return clone
+
+    def select_by(self, parent=None, child=None, dependency_types='all'):
+        """Return a list of edges selected by dependency types.
+
+        If an edge has any of the dependency types passed as argument
+        then it is selected.
+
+        Args:
+            parent (str): name of the parent package
+            child (str): name of the child package
+            dependency_types: allowed dependency types
+
+        Returns:
+            List of DependencySpec objects
+        """
+        if not dependency_types:
+            return []
+
+        dependency_types = dp.canonical_deptype(dependency_types)
+        # Start from all the edges we store
+        selected = [d for d in itertools.chain.from_iterable(self.values())]
+
+        # Filter by parent name
+        if parent:
+            selected = [d for d in selected if d.parent.name == parent]
+
+        # Filter by child name
+        if child:
+            selected = [d for d in selected if d.spec.name == child]
+
+        # Filter by allowed dependency types
+        if dependency_types:
+            selected = [
+                dep for dep in selected
+                if any(d in dependency_types for d in dep.deptypes)
+                or not dep.deptypes
+            ]
+
+        return selected
```

Comment:
Assuming all the above are now generators:
```suggestion
        return list(selected)
```

Does the result need to be a list though?

Reply:
Done in ac4a84e 

> Does the result need to be a list though?

I think returning a list of edges is sensible, and probably the simplest choice. Do you have other data structures in mind for this?

Reply:
not really -- I think this is good.